S3 driver uses wrong path for auth check.

I see in the code for the S3 driver that an initial auth check is made like this:

```
if _, err := bucket.List(strings.TrimRight(params.RootDirectory, "/"), "", "", 1); err != nil {
```
https://github.com/docker/distribution/blob/b0e94fb154789f87f0c794b5e51192548821fed3/registry/storage/driver/s3/s3.go#L224

This means that if I have specified the bucket as `testing` and root directory as `foobar` or (`foobar/`), the distribution makes a List call for the path `s3://testing/foobar`. However, since the List call in Amazon S3 is a simple prefix match, this will also return paths of the style `s3://testing/foobarbaz`. This means that coupled with a bucket policy of for example:

```
    - Effect: Allow
      Action: [ 's3:ListBucket' ]
      Resource: [ 'arn:aws:s3:::testing' ]
      Condition:
        StringLike:
          s3:prefix: ['foobar/*']
```

...causes the access check to fail as `foobar` does not match the pattern `foobar/`. S3 buckets do not behave as unix directories, so this is easy to get wrong there.

The correct fix is to always add a final slash to the "root directory" if it does not already have it, instead of stripping the slash.

Example error trace:
```
time="2015-09-18T09:45:47.46900505Z" level=info msg="debug server listening localhost:5001" 
panic: Access Denied

goroutine 1 [running]:
github.com/docker/distribution/registry/handlers.NewApp(0x7fcebe1e6f70, 0xc2080bb1d0, 0xc208063900, 0x7fcebe1e6f70)
	/go/src/github.com/docker/distribution/registry/handlers/app.go:97 +0x3e1
github.com/docker/distribution/registry.NewRegistry(0x7fcebe1e6fb0, 0xc2080bb1d0, 0xc208063900, 0xc208063900, 0x0, 0x0)
	/go/src/github.com/docker/distribution/registry/registry.go:94 +0x2d4
github.com/docker/distribution/registry.func·001(0x10ff660, 0xc20808a110, 0x1, 0x1)
	/go/src/github.com/docker/distribution/registry/registry.go:57 +0x2d1
github.com/spf13/cobra.(*Command).execute(0x10ff660, 0xc20800a010, 0x1, 0x1, 0x0, 0x0)
	/go/src/github.com/docker/distribution/Godeps/_workspace/src/github.com/spf13/cobra/command.go:495 +0x65c
github.com/spf13/cobra.(*Command).Execute(0x10ff660, 0x0, 0x0)
	/go/src/github.com/docker/distribution/Godeps/_workspace/src/github.com/spf13/cobra/command.go:560 +0x18d
main.main()
	/go/src/github.com/docker/distribution/cmd/registry/main.go:21 +0x2a

goroutine 6 [syscall]:
os/signal.loop()
	/usr/src/go/src/os/signal/signal_unix.go:21 +0x1f
created by os/signal.init·1
	/usr/src/go/src/os/signal/signal_unix.go:27 +0x35

goroutine 12 [runnable]:
net/http.(*persistConn).readLoop(0xc2080a6630)
	/usr/src/go/src/net/http/transport.go:928 +0x9ce
created by net/http.(*Transport).dialConn
	/usr/src/go/src/net/http/transport.go:660 +0xc9f

goroutine 8 [IO wait]:
net.(*pollDesc).Wait(0xc208010920, 0x72, 0x0, 0x0)
	/usr/src/go/src/net/fd_poll_runtime.go:84 +0x47
net.(*pollDesc).WaitRead(0xc208010920, 0x0, 0x0)
	/usr/src/go/src/net/fd_poll_runtime.go:89 +0x43
net.(*netFD).accept(0xc2080108c0, 0x0, 0x7fcebe1e6230, 0xc2080de568)
	/usr/src/go/src/net/fd_unix.go:419 +0x40b
net.(*TCPListener).AcceptTCP(0xc208038188, 0x4ed794, 0x0, 0x0)
	/usr/src/go/src/net/tcpsock_posix.go:234 +0x4e
net/http.tcpKeepAliveListener.Accept(0xc208038188, 0x0, 0x0, 0x0, 0x0)
	/usr/src/go/src/net/http/server.go:1976 +0x4c
net/http.(*Server).Serve(0xc2080ec840, 0x7fcebe1e8640, 0xc208038188, 0x0, 0x0)
	/usr/src/go/src/net/http/server.go:1728 +0x92
net/http.(*Server).ListenAndServe(0xc2080ec840, 0x0, 0x0)
	/usr/src/go/src/net/http/server.go:1718 +0x154
net/http.ListenAndServe(0xc20808aa90, 0xe, 0x0, 0x0, 0x0, 0x0)
	/usr/src/go/src/net/http/server.go:1808 +0xba
github.com/docker/distribution/registry.func·005(0xc20808aa90, 0xe)
	/go/src/github.com/docker/distribution/registry/registry.go:51 +0x132
created by github.com/docker/distribution/registry.func·001
	/go/src/github.com/docker/distribution/registry/registry.go:54 +0x2a7

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/src/go/src/runtime/asm_amd64.s:2232 +0x1

goroutine 13 [select]:
net/http.(*persistConn).writeLoop(0xc2080a6630)
	/usr/src/go/src/net/http/transport.go:945 +0x41d
created by net/http.(*Transport).dialConn
	/usr/src/go/src/net/http/transport.go:661 +0xcbc
```

There is no place where this even mentions S3.

--- a/registry/storage/driver/s3/s3.go
+++ b/registry/storage/driver/s3/s3.go
@@ -219,12 +219,6 @@ func New(params DriverParameters) (*Driver, error) {
 		}
 	}
 
-	// Validate that the given credentials have at least read permissions in the
-	// given bucket scope.
-	if _, err := bucket.List(strings.TrimRight(params.RootDirectory, "/"), "", "", 1); err != nil {
-		return nil, err
-	}
-
 	// TODO Currently multipart uploads have no timestamps, so this would be unwise
 	// if you initiated a new s3driver while another one is running on the same bucket.
 	// multis, _, err := bucket.ListMulti("", "")

